src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -1000,603 +946,315 @@ private ImmutableArray<Symbol> CreateMembers()
+                : this(underlyingType)
             {
-                return GetDeclaringSyntaxReferenceHelper<CSharpSyntaxNode>(_locations);
+                ElementNames = elementNames;
+                ElementLocations = elementLocations;
+                ErrorPositions = errorPositions;
+                Locations = locations.IsDefault ? ImmutableArray<Location>.Empty : locations;
             }
-        }
 
-        internal override bool Equals(TypeSymbol t2, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool> isValueTypeOverrideOpt = null)
-        {
-            if ((comparison & TypeCompareKind.IgnoreTupleNames) != 0)
+            internal bool EqualsIgnoringTupleUnderlyingType(TupleUncommonData? other)
             {
-                if (t2?.IsTupleType == true)
+                if (other is null && this.ElementNames.IsDefault && this.ElementLocations.IsDefault && this.ErrorPositions.IsDefault)
                 {
-                    t2 = t2.TupleUnderlyingType;
+                    return true;
                 }
 
-                return _underlyingType.Equals(t2, comparison, isValueTypeOverrideOpt);
-            }
+                return other is object &&
+                    this.ElementNames == other.ElementNames
+                    && compareElementLocations(other)
+                    && this.ErrorPositions == other.ErrorPositions;
 
-            return this.Equals(t2 as TupleTypeSymbol, comparison, isValueTypeOverrideOpt);
-        }
+                bool compareElementLocations(TupleUncommonData other)
+                {
+                    if (this.ElementLocations.IsDefault && other.ElementLocations.IsDefault)
+                    {
+                        return true;
+                    }
 
-        internal bool Equals(TupleTypeSymbol other)
-        {
-            return Equals(other, TypeCompareKind.ConsiderEverything);
-        }
+                    if (this.ElementLocations.IsDefault != other.ElementLocations.IsDefault)
+                    {
+                        return false;
+                    }
 
-        private bool Equals(TupleTypeSymbol other, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool> isValueTypeOverrideOpt)
-        {
-            if (ReferenceEquals(this, other))
-            {
-                return true;
+                    return this.ElementLocations.SequenceEqual(other.ElementLocations, (l1, l2) => (l1 is object && l1.Equals(l2)) || l2 is null);
+                }
             }
 
-            if ((object)other == null || !other._underlyingType.Equals(_underlyingType, comparison, isValueTypeOverrideOpt))
-            {
-                return false;
-            }
 
-            // Make sure field names are the same.
-            if ((comparison & TypeCompareKind.IgnoreTupleNames) == 0)
+            public ImmutableArray<TypeWithAnnotations> TupleElementTypesWithAnnotations(NamedTypeSymbol tuple)
```

Comment:
>tuple [](start = 104, length = 5)

What if this argument doesn't match `TupleUnderlyingType`? Either this method should take no arguments and use `_lazyElementTypes` or it should be a `static` method.

Reply:
To initialize the cache portion of TupleData (as opposed to the real data portion), we need to be handed the type (since we'll need to inspect the members to figure out the tuple elements before we cache them).

---
In reply to: [355732569](https://github.com/dotnet/roslyn/pull/39370#discussion_r355732569) [](ancestors = 355732569)

Reply:
It seems fragile to have a cached value depend on an argument. We should at least assert the argument value is expected: `Debug.Assert(tuple.TupleUnderlyingType == TupleUnderlyingType);`

---
In reply to: [355789210](https://github.com/dotnet/roslyn/pull/39370#discussion_r355789210) [](ancestors = 355789210,355732569)

Reply:
Oh, indeed. Got it :)
I forgot that we're always storing a link back to type in `TupleUncommonData` when it is created. So we no longer need to pass this parameter (redundant).

---
In reply to: [356131050](https://github.com/dotnet/roslyn/pull/39370#discussion_r356131050) [](ancestors = 356131050,355789210,355732569)

Reply:
So, it turns out that doesn't work. The tuple that is passed to these two methods is not the underlying type (which has no element names). The tuple that is passed is the `NamedTypeSymbol` with extra data for tuple elements.

---
In reply to: [356206522](https://github.com/dotnet/roslyn/pull/39370#discussion_r356206522) [](ancestors = 356206522,356131050,355789210,355732569)